lib/common.js

```diff
@@ -233,26 +229,122 @@ const headersFieldsArrayToLowerCase = function(headers) {
   )
 }
 
+/**
+ * Converts the various accepted formats of headers into a flat array representing "raw headers".
+ *
+ * Nock allows headers to be provided as a raw array, a plain object, or a Map.
+ *
+ * While all the header names are expected to be strings, the values are left intact as they can
+ * be functions, strings, or arrays of strings.
+ *
+ *  https://nodejs.org/api/http.html#http_message_rawheaders
+ */
+const headersInputToRawArray = function(headers) {
+  if (!headers) {
+    return []
+  }
+
+  if (Array.isArray(headers)) {
+    // If the input is an array, assume it's already in the raw format and simply return a copy
+    return [...headers]
+  }
+
+  // [].concat(...) is used instead of Array.flat until v11 is the minimum Node version
+  if (_.isMap(headers)) {
+    return [].concat(...Array.from(headers, ([k, v]) => [k.toString(), v]))
+  }
+
+  if (_.isPlainObject(headers)) {
+    return [].concat(...Object.entries(headers))
+  }
+
+  throw new Error(
+    `Headers must be provided as an array of raw values, a Map, or a plain Object. ${headers}`
+  )
+}
+
+/**
+ * Converts an array of raw headers to an object, using the same rules as Nodes `http.IncomingMessage.headers`.
+ *
+ * Header names/keys are lower-cased.
+ */
 const headersArrayToObject = function(rawHeaders) {
   if (!Array.isArray(rawHeaders)) {
     throw Error('Expected a header array')
   }
 
-  const headers = {}
+  const accumulator = {}
 
   for (let i = 0, len = rawHeaders.length; i < len; i = i + 2) {
-    const key = rawHeaders[i].toLowerCase()
-    const value = rawHeaders[i + 1]
+    addHeaderLine(accumulator, rawHeaders[i], rawHeaders[i + 1])
+  }
 
-    if (headers[key]) {
-      headers[key] = Array.isArray(headers[key]) ? headers[key] : [headers[key]]
-      headers[key].push(value)
+  return accumulator
+}
+
+const noDuplicatesHeaders = new Set([
+  'age',
+  'authorization',
+  'content-length',
+  'content-type',
+  'etag',
+  'expires',
+  'from',
+  'host',
+  'if-modified-since',
+  'if-unmodified-since',
+  'last-modified',
+  'location',
+  'max-forwards',
+  'proxy-authorization',
+  'referer',
+  'retry-after',
+  'user-agent',
+])
+
+/**
+ * Set key/value data in accordance with Node's logic for folding duplicate headers.
+ *
+ * https://nodejs.org/api/http.html#http_message_headers
+ * https://github.com/nodejs/node/blob/908292cf1f551c614a733d858528ffb13fb3a524/lib/_http_incoming.js#L245
+ *
+ * Header names are lower-cased.
+ * The `value` param should be a function, string, or array of strings.
+ */
+const addHeaderLine = function(headers, name, value) {
+  const key = name.toLowerCase()
+
+  let values
+  if (typeof value === 'function') {
+    // Function values are evaluated towards the end of the response, before that we use a placeholder
+    // string just to designate that the header exists. Useful when `Content-Type` is set with a function.
```

Comment:
`values` needs to be an array of strings to behave.

Reply:
Which part does? I think it's a good idea not to evaluate the function values twice, though it makes me nervous to pass around objects with half-evaluated header values.

I haven't followed exactly how this is used, though some of it seems related to checking whether other headers are _present_. Would it work to pass around the list of keys?

Reply:
Augh I see it's really involved and complicated. Would like this to be cleaner but it can be left for another day!

Reply:
To clarify, the functions are called at most once and we have a test to assert that.

`addHeaderLine` is a private helper function for `headersArrayToObject`, which is called in two places. 

First in `RequestOverrider.end`, in the case all the functions have already been evaluated so `value` will never be a function. 

The second place `headersArrayToObject` is called is in `Interceptor.reply` where headers are temporarily assembled to aid in other logic between the `Interceptor` and the `RequestOverrider`.
I left a comment there to explain why this is done: https://github.com/nock/nock/blob/89f3ebf2b91dd9889d396efb94d3a8d227c20d88/lib/interceptor.js#L112-L119

All of that is roughly the same as it was before. Previously, in `Interceptor.reply`, the headers would get grouped and the ones that happened to be functions where just left as functions in the headers object. The change I created here was caused by enforcing a parity with the headers object and what Node would create. Node's `http.IncomingMessage.headers` object  always has values of strings (expect `set-cookie` which is an array of strings). For the case of values that are functions, I had to make a decision on how to handle those given it's not a feature of Node. I didn't want to skip them because there are cases, eg `Content-Encoding`, where the existence of the field name made a difference. I chose to use the functions name as I thought it might help users when debugging. 
Passing around a list of header field names wouldn't work as there are sometimes, eg `Content-Type`, where the value in introspected. Now in those cases, if the value is a function, it's not compared correctly. But that's a shortcoming of the current code, and probably one everyone is ok living with as the alternative would be to evaluate the header functions before we have a response.